core/src/test/scala/unit/kafka/server/ServerShutdownTest.scala

```diff
@@ -195,8 +195,9 @@ class ServerShutdownTest extends KafkaServerTestHarness {
       // identify the correct exception, making sure the server was shutdown, and cleaning up if anything
       // goes wrong so that awaitShutdown doesn't hang
       case e: Exception =>
-        assertTrue(exceptionClassTag.runtimeClass.isInstance(e), s"Unexpected exception $e")
-        assertEquals(if (quorum == "zk") BrokerState.NOT_RUNNING else BrokerState.SHUTTING_DOWN, brokers.head.brokerState)
+        assertTrue(exceptionClassTag.runtimeClass.isInstance(if (isKRaftTest() && e.isInstanceOf[RuntimeException]) e.getCause.getCause else e),
```

Comment:
I'm a little confused here. You said, we just wrap a `RuntimeException` for the original exception. But here, we `getCause` twice, compared with before, it didn't `getCause`. Am I missing something? Thanks.

Reply:
Yeah, in the finally block, we will unwrap `ExecutionExecution` :
https://github.com/apache/kafka/blob/49226721c0dc5e5b327e0754e01c367990b43758/core/src/main/scala/kafka/server/BrokerServer.scala#L444

I find this is code is no longer useful after #11969, do you think we can remove it here?

Reply:
Thanks for the explanation. I understand now. So, do you think we should fix the failed tests by fixing this:

https://github.com/apache/kafka/blob/49226721c0dc5e5b327e0754e01c367990b43758/core/src/main/scala/kafka/server/BrokerServer.scala#L444

I mean, the above code is trying to throw the root cause reason the broker failed up to here:
https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/Kafka.scala#L113

And then, we can log the root cause before exit. So, I think we should fix the line `throw if (e.isInstanceOf[ExecutionException]) e.getCause else e ` to allow it to throw the root cause for both `ExecutionException` and `RuntimeException`. WDYT?

Reply:
Basically, we can change it to something like :
```
throw
if (e.isInstanceOf[ExecutionException]) 
    e.getCause
 else if (e.isInstanceOf[RuntimeException])
    e.getCause.getCause
```
this would make the logic a little messy, and I'm not sure should we should keep the original `RuntimeException` created by @cmccabe.

Reply:
Would a utility like this make sense? 
```scala
  private def assertCause(
    expectedClass: Class[_],
    e: Throwable
  ): Unit = {
    var cause = e
    while (cause != null) {
      if (expectedClass.isInstance(cause)) {
        return
      }
      cause = cause.getCause
    }
    fail(s"Failed to assert cause $expectedClass")
  }
```

Reply:
[KAFKA-13908](https://issues.apache.org/jira/browse/KAFKA-13908) is created for the RuntimeException cause issue.